### ISSUE TITLE
Docker container stuck in status (health: starting)

### ISSUE THREAD

**Issue by jackdbd** [caprover/caprover#844]
## What is the problem?

TL;DR: deploy always successful, with or without Docker `HEALTHCHECK`. But HTTP 502 error with `HEALTHCHECK`.

I have a simple Clojure web app that I dockerize with a multi-stage build (1st stage I build a .jar file, 2st stage I copy the .jar and execute it).

In my Dockerfile I have this Docker `HEALTHCHECK` instruction:

```Dockefile
HEALTHCHECK --interval=5m --timeout=3s CMD curl --fail https://localhost/ || exit 1
```

Whenever I deploy the app on CapRover with this `HEALTHCHECK`, the deployment is successful but I get a HTTP 502 error from nginx if I visit the page. The app runs on port 3000, but that port is correctly exposed in the `Dockerfile` and configured in the HTTPS Settings of the CapRover UI.

If I access my VPS (a DigitalOcean droplet) via SSH and run `docker ps` I see that a container is stuck in a status of `(healthy: starting)`. That's my clojure app.

![image](https://user-images.githubusercontent.com/5048090/94855233-7aca0a00-042e-11eb-8dbf-6b3f17eef74c.png)

If I stop the container, run `docker system prune` and redeploy without the `HEALTHCHECK`, the app runs as expected.

---

When I googled for possible reasons for this behavior I found this sentence from [this issue](https://github.com/moby/moby/issues/37067):

> docker is stuck querying the status of a healthcheck which brings a whole bunch of things to a halt.

So I wondered... is this a Docker Swarm issue?

## My `captain-definition`

```json
{
  "schemaVersion": 2,
  "dockerfilePath": "./Dockerfile"
}
```

# My `Dockerfile`

```Dockerfile
# === STAGE 1 ================================================================ #
# Build a standalone .jar file with Leiningen
# ============================================================================ #
FROM clojure:latest AS builder

ENV APP_DIR=/usr/src/app \
    JAR_FILE=tweedler-standalone.jar

RUN mkdir -p ${APP_DIR}

WORKDIR ${APP_DIR}

COPY project.clj ${APP_DIR}/
COPY resources ${APP_DIR}/resources
COPY src ${APP_DIR}/src

# The Clojure project.clj includes the lein-ring plugin, so we can use this
# command to build the .jar
# https://github.com/weavejester/lein-ring
RUN lein ring uberjar

# === STAGE 2 ================================================================ #
# Copy the .jar built at stage 1 and execute it
# ============================================================================ #

# I think we can safely run the generated .jar on a JVM which runs on a
# different Linux distro. For this second stage we use Alpine Linux instead of
# clojure:latest (which is based on Debian). This way we save a few MB on
# the Docker image.
FROM openjdk:8-jre-alpine3.9

ENV SRC_DIR=/usr/src/app \
    USER_HOME=/home/appuser \
    JAR_FILE=tweedler-standalone.jar \
    PORT=3000 \
    JVM_OPTS=-Dclojure.main.report=stderr

# Create a group and user
RUN addgroup -S appgroup && adduser -S appuser -G appgroup

USER appuser

WORKDIR ${USER_HOME}

COPY --from=builder ${SRC_DIR}/target/uberjar/${JAR_FILE} ${USER_HOME}/${JAR_FILE}

EXPOSE ${PORT}

# Check every 5 minutes that a web-server is able to serve the site's main page
# within 3 seconds.
# https://docs.docker.com/engine/reference/builder/#healthcheck
# I think I can't use a Docker HEALTHCHECK on CapRover. It seems it causes the
# container to hang in a status of: `(health: starting)`
# HEALTHCHECK --interval=5m --timeout=3s \
#   CMD curl --fail https://localhost/ || exit 1
 
CMD java -jar ${JVM_OPTS} ${USER_HOME}/${JAR_FILE}
```

## My environment

CapRover runs on a 5 USD DigitalOcean droplet (1GB RAM, 1 CPU). There is only CapRover on this machine.

The OS on my laptop is Xubuntu 20.04.1 LTS (16 GB RAM).

**Comment by githubsaturn** [comment]
Is this a CapRover issue? It looks like a Docker issue.

Does it run fine with the same Dockerfile if you just use `docker run imagename` ?

**Comment by jackdbd** [comment]
Yes, it runs without any issue if I run `docker run -p 3000:3000 imagename` on my laptop.

**Comment by githubsaturn** [comment]
Interesting. Does it also work if you map to another port other than 3000? Like
```
docker run -p 3001:3000 imagename
```

In your healthcheck you're not defining a port, also it's `https` - it'd be good to try `https://localhost:3000/` or  `https://127.0.0.1:3000/`

**Comment by jackdbd** [comment]
My bad. The reason was that I was using curl, which is not installed by default in Alpine images.

I replaced curl with wget in my HEALTHCHECK and the container soon changed to `healthy`. I also had to specify the port and use http instead of https.

Here is my HEALTHCHECK with wget:

```Dockerfile
HEALTHCHECK --interval=1m --timeout=3s --retries=3 \
  CMD wget --no-verbose --tries=1 --spider http://localhost:${PORT}/ || exit 1
```

I also realized only now that with my old curl HEALTHCHECK the container on my laptop became `unhealthy` after a few minutes. That didn't seem to affect the application though.

**Comment by Sahilbhawke** [comment]
> My bad. The reason was that I was using curl, which is not installed by default in Alpine images.
> 
> I replaced curl with wget in my HEALTHCHECK and the container soon changed to `healthy`. I also had to specify the port and use http instead of https.
> 
> Here is my HEALTHCHECK with wget:
> 
> ```dockerfile
> HEALTHCHECK --interval=1m --timeout=3s --retries=3 \
>   CMD wget --no-verbose --tries=1 --spider http://localhost:${PORT}/ || exit 1
> ```
> 
> I also realized only now that with my old curl HEALTHCHECK the container on my laptop became `unhealthy` after a few minutes. That didn't seem to affect the application though.

Can you tell how much time it took to the "healthy" state?

**Comment by F8ER** [comment]
> Can you tell how much time it took to the "healthy" state?

```
HEALTHCHECK...
 --health-start-period   Start period for the container to initialize before starting health-retries countdown
```
Source: <https://docs.docker.com/engine/reference/run/#healthcheck>
